<div dir="rtl">
    
###

   - **יש למלא את המחברת באנגלית בלבד**
   - **מומלץ להשתמש בהמשך באותו הקובץ להגשת הפרויקט (מיחקו את התא הזה)**
   - **הקפידו למלא טקסט בשדות מסוג מארק-דאון וקוד בשדות מסוג קוד (אחרת ירדו לכם נקודות)**
     
###   המחברת תכלול  (לכל הפחות):
- שם הפרויקט
- שמות מלאים של כל חברי הצוות, ללא ת.ז
- תיאור הפרויקט (3-4 משפטים)
- קישור למקור ממנו הורדתם את הנתונים  
- הסבר על מקור הנתונים
- תיאור קצר של הדאטה בעזרת הפקודות שנלמדו בקורס ונעשה בהם שימוש בפרוירט
- שלוש שאלות ראשוניות של דברים שהיה מעניין אתכם להוציא מהקובץ
#### בהצלחה !
</div>

In [50]:
#שאלה ראשונה: האם צריך רק תז או שם מלא  
#האם מותר לשנות שם עמודה בדאתא כדי שנזהה אותה נכון (אני מדברת על עמודה # שהיא עמודת מזהה ואנחנו רוצות לקרוא לה בשם אחר)

# Project Name: `Layoff Trends in the Global Tech Industry (2020–2024)`


### Project team members:
- `מרים דינאי`
- `רוני שפירא`

### Project number: <insert here>
`10`

### Project description:
`1-3 lines describing your project`

### Link to the data:
The dataset was downloaded from Kaggle:  
https://www.kaggle.com/datasets/ulrikeherold/tech-layoffs-2020-2024

### Explain the data source:
`1-3 lines describing how this data was gathered, where from etc.`

----

#### Read you file from Github:

In [59]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

url='https://raw.githubusercontent.com/MiriamDinay/Layoff-Trends/main/tech_layoffs.csv'
Layoff_df=pd.read_csv(url)
Layoff_df.head()

#         Company             Location_HQ Country      Continent  \
0   3       ShareChat               Bengaluru   India           Asia   
1   4       InSightec                   Haifa  Israel           Asia   
2   6  Enphase Energy  San Francisco Bay Area     USA  North America   
3   7           Udaan               Bengaluru   India           Asia   
4  14          Cruise  San Francisco Bay Area     USA  North America   

   Laid_Off Date_layoffs  Percentage  Company_Size_before_Layoffs  \
0     200.0   2023-12-20        15.0                       1333.0   
1     100.0   2023-12-19        20.0                        500.0   
2     350.0   2023-12-18        10.0                       3500.0   
3     100.0   2023-12-18        10.0                       1000.0   
4     900.0   2023-12-14        24.0                       3750.0   

   Company_Size_after_layoffs        Industry     Stage Money_Raised_in_$_mil  \
0                      1133.0        Consumer  Series H                 $1700   
1                       400.0      Healthcare   Unknown                  $733   
2                      3150.0          Energy  Post-IPO                  $116   
3                       900.0          Retail   Unknown                   NaN   
4                      2850.0  Transportation  Acquired                $15000   

   Year       lat        lng  
0  2023  12.97194   77.59369  
1  2023  32.81841   34.98850  
2  2023  37.54827 -121.98857  
3  2023  12.97194   77.59369  
4  2023  37.77493 -122.41942

#### Describe the data
* How many rows?
* How many columns?
* How many numeric columns? (not including numerical identifiers - numbers that uniquely define the row)
* How many categorical columns? (not including string identifiers -  values that uniquely define the row)

In [51]:
#How many rows?
#How many columns?

Layoff_df.shape

(1672, 16)

In [60]:
Layoff_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1672 entries, 0 to 1671
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   #                            1672 non-null   int64  
 1   Company                      1672 non-null   object 
 2   Location_HQ                  1672 non-null   object 
 3   Country                      1672 non-null   object 
 4   Continent                    1672 non-null   object 
 5   Laid_Off                     1565 non-null   float64
 6   Date_layoffs                 1672 non-null   object 
 7   Percentage                   1570 non-null   float64
 8   Company_Size_before_Layoffs  1511 non-null   float64
 9   Company_Size_after_layoffs   1536 non-null   float64
 10  Industry                     1672 non-null   object 
 11  Stage                        1672 non-null   object 
 12  Money_Raised_in_$_mil        1596 non-null   object 
 13  Year              

In [86]:
# לשים לב לכתוב על מחיקת השורות שאנחנו לא צריכים למחקר. אחר כך נרשום הכל בצורה מסודרת
Layoff_df_copy = Layoff_df.copy()
Layoff_df_copy['Date_layoffs'] = pd.to_datetime(Layoff_df_copy['Date_layoffs'])
Layoff_df_copy['Money_Raised_in_$_mil'] = (Layoff_df_copy['Money_Raised_in_$_mil'].astype(str).str.replace(r'[$,]', '', regex=True))
Layoff_df_copy['Money_Raised_in_$_mil'] = pd.to_numeric(Layoff_df_copy['Money_Raised_in_$_mil'],errors='coerce')
Layoff_df_copy['#'] = Layoff_df_copy['#'].astype(str)
Layoff_df_copy = Layoff_df_copy.drop(columns=['lat', 'lng',])
Layoff_df_copy.dtypes

#                                      object
Company                                object
Location_HQ                            object
Country                                object
Continent                              object
Laid_Off                              float64
Date_layoffs                   datetime64[ns]
Percentage                            float64
Company_Size_before_Layoffs           float64
Company_Size_after_layoffs            float64
Industry                               object
Stage                                  object
Money_Raised_in_$_mil                 float64
Year                                    int64
dtype: object

In [87]:
# בדיקה אם יש הבדלים בשנה והאם אפשר לסמוך על העמודה הזאת
mismatch = Layoff_df_copy[Layoff_df_copy['Date_layoffs'].dt.year != Layoff_df_copy['Year']]
mismatch.shape

(37, 14)

בבדיקת עקביות נוספת, נמצאו 37 רשומות שבהן שנת הפיטורין המדווחת לא תאמה את התאריך המלא. כדי להבטיח דיוק בניתוח הנתונים, החלטנו להתבסס אך ורק על עמודת ה-Date_layoffs ולעדכן את עמודת השנה לפיה.

In [91]:
# דריסת עמודת השנה המקורית בשנה המדויקת מהתאריך
Layoff_df_copy['Year'] = Layoff_df_copy['Date_layoffs'].dt.year

In [73]:
Layoff_df_copy.isnull().sum().sort_values(ascending=False)

Company_Size_before_Layoffs    161
Company_Size_after_layoffs     136
Laid_Off                       107
Percentage                     102
Money_Raised_in_$_mil           76
#                                0
Continent                        0
Country                          0
Location_HQ                      0
Company                          0
Date_layoffs                     0
Industry                         0
Stage                            0
Year                             0
dtype: int64

בחרתי לבצע המרת טיפוסי נתונים לפני ניקוי שורות כדי לשמר את שלמות בסיס הנתונים המקורי. ערכים חסרים לא הושלמו באופן מלאכותי כדי למנוע הטיה בניתוח חומרת הפיטורין וגודל החברות

In [74]:
# 4. ניקוי שורות שאין בהן מידע על פיטורין בכלל (כמות או אחוז)
# זהו השלב שבו אנחנו מוודאים שכל שורה ב-Dataset יכולה לתרום למחקר
Layoff_df_copy = Layoff_df_copy.dropna(subset=['Laid_Off', 'Percentage'], how='all')

# 5. בקרת איכות (Sanity Check) - השוואת הפרשי גדלים לכמות המפוטרים
# הבדיקה מתבצעת רק היכן שכל שלושת הנתונים קיימים במקור
mask = (
    Layoff_df_copy['Company_Size_before_Layoffs'].notna() &
    Layoff_df_copy['Company_Size_after_layoffs'].notna() &
    Layoff_df_copy['Laid_Off'].notna()
)

Layoff_df_copy.loc[mask, 'size_diff_check'] = (
    Layoff_df_copy['Company_Size_before_Layoffs'] -
    Layoff_df_copy['Company_Size_after_layoffs'] -
    Layoff_df_copy['Laid_Off']
)

# הצגת סיכום הנתונים לאחר הניקוי
print(f"Total rows after cleaning: {len(Layoff_df_copy)}")
print(Layoff_df_copy.dtypes)

Total rows after cleaning: 1624
#                                      object
Company                                object
Location_HQ                            object
Country                                object
Continent                              object
Laid_Off                              float64
Date_layoffs                   datetime64[ns]
Percentage                            float64
Company_Size_before_Layoffs           float64
Company_Size_after_layoffs            float64
Industry                               object
Stage                                  object
Money_Raised_in_$_mil                 float64
Year                                    int64
size_diff_check                       float64
dtype: object


In [75]:
inconsistent_rows = Layoff_df_copy[Layoff_df_copy['size_diff_check'] != 0].shape[0]
print(f"Number of rows with inconsistencies: {inconsistent_rows}")

Number of rows with inconsistencies: 114


הפער שנמצא ב-114 רשומות מדגיש ששינויים במצבת כוח האדם מושפעים מגורמים נוספים מלבד פיטורים מוכרזים, כגון עזיבה מרצון או שינויים ארגוניים פנימיים. לכן, שמירה על הנתונים המקוריים מבטיחה ניתוח אובייקטיבי יותר

In [76]:
(Layoff_df_copy.loc[mask, 'size_diff_check'] == 0).mean()

np.float64(0.99933818663137)

כדי להעריך את איכות הנתונים, ביצענו בדיקת הצלבה (Sanity Check) על תת-קבוצה של רשומות המכילות נתונים מלאים לגבי גודל החברה (לפני ואחרי) ומספר המפוטרים.

ממצאי הבדיקה:

מתוך השורות שנבדקו (ה-mask), נמצא כי ב-114 רשומות קיים פער מתמטי.

פער זה אינו מעיד על חוסר אמינות של ה-Dataset, אלא על כך שהשינוי במצבת כוח האדם מושפע מגורמים נוספים כמו עזיבה מרצון או גיוס במקביל.

בדיקה זו חיזקה את ההחלטה המתודולוגית לא לבצע השלמת נתונים (Imputation) בצורה אוטומטית, אלא לעבוד עם הנתונים המקוריים בלבד כדי לשמור על אובייקטיביות הניתוח

##### יצרנו העתק של קובץ הנתונים המקורי, סידרנו אותם וסיווגנו את "טיפוסי" העמודות.
##### מעתה, בהמשך העבודה נעבוד על קובץ ההעתק בלבד.

In [78]:
Layoff_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1624 entries, 0 to 1671
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   #                            1624 non-null   object        
 1   Company                      1624 non-null   object        
 2   Location_HQ                  1624 non-null   object        
 3   Country                      1624 non-null   object        
 4   Continent                    1624 non-null   object        
 5   Laid_Off                     1565 non-null   float64       
 6   Date_layoffs                 1624 non-null   datetime64[ns]
 7   Percentage                   1570 non-null   float64       
 8   Company_Size_before_Layoffs  1511 non-null   float64       
 9   Company_Size_after_layoffs   1536 non-null   float64       
 10  Industry                     1624 non-null   object        
 11  Stage                        1624 non-null   obj

In [80]:
#האם כדאי להשאיר או למחוק את עמודות הקורדינטות והסולמית.
# Layoff_df_copy.describe()

In [112]:
# How many numeric columns?
numeric_cols = Layoff_df_copy.select_dtypes(include=['int64', 'float64']).columns

print(f"Number of numerical columns: {len(numeric_cols)}")
print(f"The columns are: {list(numeric_cols)}")

Number of numerical columns: 5
The columns are: ['Laid_Off', 'Percentage', 'Company_Size_before_Layoffs', 'Company_Size_after_layoffs', 'Money_Raised_in_$_mil']


In [110]:
# How many categorical columns?
categorical_cols = Layoff_df_copy.select_dtypes(include=['object']).columns
categorical_cols = categorical_cols.drop(['Company', '#']) #האם להוריד או להשאיר?

print(f"Number of meaningful categorical columns: {len(categorical_cols)}")
print(f"The columns are: {list(categorical_cols)}")

Number of meaningful categorical columns: 5
The columns are: ['Location_HQ', 'Country', 'Continent', 'Industry', 'Stage']


Categorical columns were identified based on object data types, while string identifiers such as the company name were excluded from the analysis.

#### Write three questions you would like to answer using this data

1. `How do layoff patterns vary over time, and are there noticeable peaks during specific periods between 2020 and 2024?`

2. `How does company stage influence the frequency and severity of layoffs in the tech industry?`

3. `Do smaller companies experience more severe layoffs relative to their size compared to larger companies?`

In [48]:
#האם אפשרי לעשות 4 שאלות?

<div style="background-color: #f0f0f0; padding: 10px; border-radius: 5px;">
  <span style="color: red;"><strong>— Do not proceed beyond this point. This is for the next steps, do not fill in —</strong></span>
</div>


---
### <span style="color:green"> Observation 1:</span>
How do layoff patterns vary over time, and are there noticeable peaks during specific periods between 2020 and 2024?
>
>
>
>


---

---
### <span style="color:green"> Observation 2:</span>

>
>
>
>


---

---
### <span style="color:green"> Observation 3:</span>

>
>
>
>


---

---
### <span style="color:green"> Summary</span>

> Anything you'd like to say - your conclusions, summary, how hard you worked etc.
>
>    
>   
   

---